finesse/hardware/dp9800.py

```diff
@@ -0,0 +1,242 @@
+
+        The system flag is stored as a bit mask with the format TxxLxSAF,
+        where:
+            F - bit 0: temperature unit: 0 = deg C, 1 = deg F
+            A - bit 1: audible button:   0 = silence, 1 = audible
+            S - bit 2: autoscan:         0 = no scan, 1 = autoscan active
+            x - bit 3: must be 0
+            L - bit 4: logging:          0 = no logging, 1 = logging active
+            x - bit 5: must be 0
+            x - bit 6: must be 0
+            T - bit 7: instrument type:  0 = TC, 1 = PT
+        """
+        if self._sysflag == "":
+            print("No system flag. Read from device first.")
+        else:
+            instr_type = ["TC", "PT"][int(self._sysflag[0])]
+            logging_state = ["no logging", "logging active"][int(self._sysflag[3])]
+            scanning_state = ["no scan", "autoscan active"][int(self._sysflag[5])]
+            audible_state = ["silence", "audible"][int(self._sysflag[6])]
+            temp_unit = ["deg C", "deg F"][int(self._sysflag[7])]
+            print(f"Instrument type: {instr_type}")
+            print(f"Logging: {logging_state}")
+            print(f"Autoscan: {scanning_state}")
+            print(f"Audible button: {audible_state}")
+            print(f"Temperature unit: {temp_unit}")
+
+    def read(self) -> bytes:
+        """Read temperature data from the DP9800.
+
+        The DP9800 returns a sequence of bytes containing the
+        temperatures measured on each channel, in the format
+
+          STX T SP t1 SP t2 SP t3 SP t4 SP t5 SP t6 SP t7 SP t8 SP t9 ff ETX BCC
+
+        where
+            t1, t2, ..., t9: temperature values in the format %4.2f
+            STX: Start of Text (ASCII 2)
+            ETX: End of Text (ASCII 3)
+            SP: Space (ASCII 32)
+            BCC: Block Check Character
+            ff: System flag in hexadecimal
+
+        For error checking, the BCC is calculated by performing consecutive XOR
+        operations on the message and compared with the BCC received.
+
+        Note a slight peculiarity:
+        The device actually returns 9 temperatures, while the documentation
+        states that there should be 8. The device shows 8 values, corresponding
+        to the values with indices 1 to 8. The first value (at index 0) is
+        therefore ignored.
+
+        Returns:
+            data: the sequence of bytes read from the device
+        """
+        num_bytes_to_read = self.serial.in_waiting
+        if num_bytes_to_read == 0:
+            data = b""
+        else:
+            try:
+                data = self.serial.read(num_bytes_to_read)
+            except SerialException as e:
+                self._error_occurred(DP9800Error(e))
+
+            # Perform message integrity checks
+            # Check characters we know
+            if data[0] != 2:  # STX
+                self._error_occurred(
+                    DP9800Error("Start transmission character not detected")
+                )
```

Comment:
I think it will be more readable if each of this checks is a different function.